### ISSUE TITLE
Can you read from multiple sources kafka clusters and write to multiple kafka clusters?

### ISSUE THREAD

**Issue by murtazapitalwala** [uber/uReplicator#158]
Hello,

We are doing a POC for uReplicator to be used across multiple regions/sites. Let me share the configs:

```
------- Controller Configs ------- 
cat start-controller-with-args.sh 
#!/bin/bash

export hoa_srcCluster='ZK-CLUSTER-1:2181/srcCluster'
export asd_desCluster='ZK-CLUSTER-2/destCluster'
export hoa_zk='ZK-CLUSTER-1:2181'

args="-helixClusterName testMirrorMaker"
args="${args} -destKafkaZkPath ${asd_desCluster}"
args="${args} -srcKafkaZkPath ${hoa_srcCluster}"
args="${args} -zookeeper ${hoa_srcCluster}"
args="${args} -port 10000 -mode auto"
args="${args} -enableAutoWhitelist true"
args="${args} -autoRebalanceDelayInSeconds 120 -backUpToGit false"
args="${args} -localBackupFilePath ~/backup_"
echo " /bin/bash ./start-controller.sh startMirrorMakerController ${args}"
/bin/bash ./start-controller.sh startMirrorMakerController ${args}

------- Worker configs ------- 
cat /app/uReplicator/config/consumer.properties  | grep -vE "^#"
zookeeper.connect=ZK-CLUSTER-1:2181/srcCluster
zookeeper.connection.timeout.ms=30000
zookeeper.session.timeout.ms=30000
group.id=kloak-mirrormaker-test
consumer.id=kloakmms01-sjc1
partition.assignment.strategy=roundrobin
socket.receive.buffer.bytes=1048576
fetch.message.max.bytes=8388608
queued.max.message.chunks=5
auto.offset.reset=smallest


cat /app/uReplicator/config/producer.properties  | grep -vE '^#'
bootstrap.servers=KAFKA-CLUSTER-2:9092
client.id=kloak-mirrormaker-test
producer.type=async
compression.type=none
serializer.class=kafka.serializer.DefaultEncoder
key.serializer=org.apache.kafka.common.serialization.ByteArraySerializer
value.serializer=org.apache.kafka.common.serialization.ByteArraySerializer
```

We would like to aggregate the data in multiple kafka cluster/topics: Can we read from 2 clusters (Cluster1 and Cluster2) and write to both clusters in different topics? Or, do you suggest read from one cluster (cluster1) and write to two clusters (cluster1 and cluster2) in different topics?

from looking at the config, I am not sure if thats possible. 

For this scenario what is the best practices/solution?

**Comment by xhl1988** [comment]
@murtazapitalwala One uReplicator cluster can only copy from one cluster to another cluster. We have similar set up in production (but using the Federated uReplicator branch). For example:
To replicator topic A from cluster1 to topic A_other_cluster in cluster2, you need to set up one cluster with --topicmapping.properties. For other direction, you need to set up another cluster. 
Regarding the reading from 1 or 2 cluster, it depends on your use cases.

**Comment by murtazapitalwala** [comment]
Hello @xhl1988, thanks for quick reply. 

I was able to copy data from cluster 1 topic_1 to cluster 2 topic_2 from the master branch of this repo.

Can One uReplicator cluster read from multiple kafka clusters (cluster1 and cluster2) and produce to local kafka cluster1? If yes, how?

Thanks,
Murtaza

**Comment by murtazapitalwala** [comment]
@xhl1988, 

Cluster 1 (AWS EAST) and Cluster 2 (AWS WEST )have the same Topics and same replication_factor.

If the data comes to cluster 1 Topic A, it should be sent to cluster 2 Topic A. And, vice versa.

If we put one uReplicator cluster to listen from cluster 1 and send to cluster 2 and another uReplicator cluster to listen from cluster 2 and send to cluster 1, it could/will create a infinite loop. 

Therefore, the reason for asking if there is a better way of doing this?

Thanks,
Murtaza

**Comment by xhl1988** [comment]
If you want to aggregate the data between two clusters, there are several use cases I can think of now:
1. (Based on your description, seems this is not your case, but I still want to call out as this is what we are doing now) 
  If your message can be filtered based on src cluster info inside message body/header, you can aggregate the message into the same topic name in the other cluster.
pros: save disk space since only two copies for each message.
cons: consumer need to do filter if they only want local data

2. If your message cannot be filtered, first solution is to only copy topics to other cluster with a postfix/prefix, e.g., cluster1 has A, B_other_cluster, cluster2 has B, A_other_cluster
pros: save disk space
cons: consumer failover logic might be more complicated

3. second solution is to copy all the topics even to it's own cluster, e.g., cluster1 has A, A_agg, B_agg, cluster2 has B, A_agg, B_agg.
pros: consumer has the flexibility to choose where to read from
cons: double disk space and more replication pipeline to maintain in case your have more clusters in the future.

So it really depends on your consumer all-active story and resource limitation. Especially for critical use cases, 1/2 might be better since there is one less hop.

**Comment by murtazapitalwala** [comment]
@xhl1988,

Thanks very much for this detailed reply. 

If we want to move forward with option 2, does this mean we need two ureplicator clusters? The first one for cluster1 and the second *separate* cluster for cluster 2? And, this is using Federated uReplicator branch?

Thanks,
Murtaza

**Comment by xhl1988** [comment]
Yes. You need two clusters. However, since you only have two cluster, I would suggest just use the master branch. (We will merge the Federated uReplicator branch into master in the future as it is compatible).

**Comment by murtazapitalwala** [comment]
@xhl1988,

Thanks for all the information. We will try that out and update the results here. 

Thanks, 
Murtaza

**Comment by murtazapitalwala** [comment]
@xhl1988,

It seems as we want to aggregate on both clusters. So let me explain the scenario.

We have two clusters: Cluster 1 and Cluster 2 with same topic A contains partial data from both clusters. To aggregate the data from both clusters:

We want to read from Cluster 1 and write to topic B_aggre in Cluster 1 and Cluster 2, read from Cluster 2 and write to topic B_aggre in Cluster 1 and Cluster 2.

So, is it possible to write to two clusters?

Thanks,
Murtaza

**Comment by xhl1988** [comment]
Federated-uReplicator has a federation layer on topic of master branch. It automatically set up cluster1->cluster2 and cluster2->cluster1 pipelines for you. However, which pipeline can only have one src and one dst cluster.
e.g. In Federated-uReplicator branch, you need to set up one manager cluster, one controller cluster, one worker cluster, but cluster1->cluster2 and cluster2->cluster1 replication will be handled by manager.
In master branch, you need to set up two controller cluster, two worker cluster, one handles cluster1->cluster2 and the other handles cluster2->cluster1.

**Comment by xhl1988** [comment]
If you need more info on Federated uReplicator, you can refer to this [meetup talk](https://www.youtube.com/watch?v=T2RDH0v3pzs), starting at around 17:00.

**Comment by murtazapitalwala** [comment]
@xhl1988,

Thanks for the info above. However, I am not sure if the answer to below questions were provided.

```
We want to read from Cluster 1 and write to topic B_aggre in Cluster 1 and Cluster 2, read from Cluster 2 and write to topic B_aggre in Cluster 1 and Cluster 2.

-- OR --

Read from Cluster 1 and Cluster 2 topic A, write to Cluster 1 topic B_aggre, and Read from Cluster 1 and Cluster 2 Topic A, write to Cluster 2 topic B_aggre
```

You did say
> However, which pipeline can only have one src and one dst cluster.

Does that mean we **CAN NOT** read from two clusters or write to two clusters from one urep cluster (controller and workers) from master or Federated-uReplicator branch?

Thanks,
Murtaza

**Comment by xhl1988** [comment]
Sorry for the confusion. Let's see if this is more clear: one urep cluster can only handle one src and one dst cluster. In your case, since you only have two clusters, you won't get much benefit from Federated-uReplicator branch:
In master branch, you need to set up two urep clusters yourself.
In Federated-uReplicator branch, you still to set up two federation cluster (one in each region), then the federation layer will automatically create one urep clusters in each region for you when you whitelist the topic. 

However, in the following case, Federated-uReplicator branch will gain you much benefit:
You have cluster1,2 in US east and cluster 3,4 in US west.
In master brach, you need to manually create 12 clusters yourself.
In Federated-uReplicator branch, you still only need to two federation cluster (one in each region). all those 12 cluster will be created automatically for you.

**Comment by murtazapitalwala** [comment]
@xhl1988,

Thanks for all the info. I really appreciate that.

Thanks,
Murtaza
